src/controllers/gambling/index.ts

```diff
@@ -177,4 +177,175 @@ export const gamblingRouter = async function (app: FastifyInstance) {
+					additionalProperties: false,
+					properties: {
+						server_id: { type: "string" },
+						user_id: { type: "string" },
+						mentioned_user_id: { type: "string" },
+						wager: { type: "number", default: 0 }
+					}
+				}
+			}
+		},
+		async (req) => {
+			const { server_id, user_id, mentioned_user_id, wager } = req.body;
+			await servers.assertExists({ server_id });
+
+			const [user, mentionedUser] = await Promise.all([
+				users.assertRead({ user_id, server_id }),
+				users.assertRead({ user_id: mentioned_user_id, server_id })
+			]);
+
+			if (user.user_id === mentionedUser.user_id)
+				throw new BadRequestError("You cannot play Connect Four against yourself!");
+
+			const [
+				existingChallenge,
+				userHasActiveGame,
+				mentionedUserHasActiveGame,
+				userHasEnoughBucks,
+				mentionedUserHasEnoughBucks
+			] = await Promise.all([
+				connectFourGames.getUnacceptedChallenge(user),
+				connectFourGames.hasActiveGame(user),
+				connectFourGames.hasActiveGame(mentionedUser),
+				users.hasBucks(user.user_id, server_id, wager),
+				users.hasBucks(mentionedUser.user_id, server_id, wager)
+			]);
+
+			if (userHasActiveGame)
+				throw new BadRequestError("You already have an active game of Connect Four!");
+			if (mentionedUserHasActiveGame)
+				throw new BadRequestError(
+					`<@${mentionedUser.user_id}> already has an active game of Connect Four!`
+				);
+
+			if (wager < 0) throw new BadRequestError("Wager amount must not be negative!");
+			if (!userHasEnoughBucks)
+				throw new BadRequestError(
+					`You only have ${user.billy_bucks} BillyBucks - cannot wager ${wager} BillyBucks!`
+				);
+			if (!mentionedUserHasEnoughBucks)
+				throw new BadRequestError(
+					`<@${mentionedUser.user_id}> only has ${mentionedUser.billy_bucks} BillyBucks - cannot wager ${wager} BillyBucks!`
+				);
+
+			if (existingChallenge) {
+				if (existingChallenge.yellow_user_id === user.user_id) {
+					// if the user is yellow in the existing challenge, check if the mentioned user is red
+					if (existingChallenge.red_user_id === mentionedUser.user_id) {
+						// start the game
+						const [startNewGame] = await Promise.all([
+							connectFourGames.acceptExistingChallengeAndStartGame(existingChallenge),
+							users.updateOne({ _id: user._id }, { $inc: { billy_bucks: -wager } }),
+							users.updateOne(
+								{ _id: mentionedUser._id },
+								{ $inc: { billy_bucks: -wager } }
+							)
+						]);
+						return startNewGame;
+					}
+					// issue a new challenge to the mentioned user
+					return await connectFourGames.createNewChallenge(user, mentionedUser, wager);
+				}
+				// update the existing challenge
+				return await connectFourGames.updateExistingChallenge(
+					existingChallenge,
+					mentionedUser,
+					wager
+				);
+			}
+			// issue a new challenge to the mentioned user
+			return await connectFourGames.createNewChallenge(user, mentionedUser, wager);
```

Comment:
swap this logic so it returns early first, the if outer statement is too big

Reply:
Refactored this whole section, should be better now.